File: ChartsDemo/Classes/RealmDemos/RealmBarChartViewController.m
//
//  RealmBarChartViewController.m
//  ChartsDemo
//
//  Copyright 2015 Daniel Cohen Gindi & Philipp Jahoda
//  A port of MPAndroidChart for iOS
//  Licensed under Apache License 2.0
//
//  https://github.com/danielgindi/Charts
//

#import "RealmBarChartViewController.h"
#import "ChartsDemo-Swift.h"
#import <Realm/Realm.h>
#import "RealmDemoData.h"

@interface RealmBarChartViewController () <ChartViewDelegate>

@property (nonatomic, strong) IBOutlet BarChartView *chartView;

@end

@implementation RealmBarChartViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [self writeRandomDataToDbWithObjectCount:20];
    
    self.title = @"Realm.io Bar Chart";
    
    self.options = @[
                     @{@"key": @"toggleValues", @"label": @"Toggle Values"},
                     @{@"key": @"toggleHighlight", @"label": @"Toggle Highlight"},
                     @{@"key": @"animateX", @"label": @"Animate X"},
                     @{@"key": @"animateY", @"label": @"Animate Y"},
                     @{@"key": @"animateXY", @"label": @"Animate XY"},
                     @{@"key": @"saveToGallery", @"label": @"Save to Camera Roll"},
                     @{@"key": @"togglePinchZoom", @"label": @"Toggle PinchZoom"},
                     @{@"key": @"toggleAutoScaleMinMax", @"label": @"Toggle auto scale min/max"},
                     ];
    
    _chartView.delegate = self;
    
    [self setupBarLineChartView:_chartView];
    
    [self setData];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)setData
{
    
    RLMRealm *realm = [RLMRealm defaultRealm];
    
    RLMResults *results = [RealmDemoData allObjectsInRealm:realm];
    
    RealmBarDataSet *set = [[RealmBarDataSet alloc] initWithResults:results xValueField:@"xValue" yValueField:@"yValue"];
    set.colors = @[[ChartColorTemplates colorFromString:@"#FF5722"],
                   [ChartColorTemplates colorFromString:@"#03A9F4"]];
    set.label = @"Realm BarDataSet";

    NSArray<id <IChartDataSet>> *dataSets = @[set];

    BarChartData *data = [[BarChartData alloc] initWithDataSets:dataSets];
    [self styleData:data];
    
    [_chartView zoomWithScaleX:5.f scaleY:1.f x:0.f y:0.f];
    _chartView.fitBars = YES;
    _chartView.data = data;
    
    [_chartView animateWithYAxisDuration:1.4 easingOption:ChartEasingOptionEaseInOutQuart];
}

- (void)optionTapped:(NSString *)key
{
    [super handleOption:key forChartView:_chartView];
}

#pragma mark - ChartViewDelegate

- (void)chartValueSelected:(ChartViewBase * __nonnull)chartView entry:(ChartDataEntry * __nonnull)entry highlight:(ChartHighlight * __nonnull)highlight
{
    NSLog(@"chartValueSelected");
}

- (void)chartValueNothingSelected:(ChartViewBase * __nonnull)chartView
{
    NSLog(@"chartValueNothingSelected");
}

@end

File: ChartsDemo-OSX/ChartsDemo-OSX/RealmDemo/RealmBarChartViewController.swift
//
//  RealmBarChartViewController.swift
//  ChartsDemo-OSX
//
//  Created by thierryH24A on 02/04/2017.
//  Copyright © 2017 dcg. All rights reserved.
//

import Cocoa
import Charts
import Realm
import ChartsRealm

class RealmBarChartViewController: RealmDemoBaseViewController {
    
    
    @IBOutlet var chartView: BarChartView!
    
    override open func viewDidAppear()
    {
        super.viewDidAppear()
        view.window!.title = "Realm.io Bar Chart"
    }

    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        writeRandomDataToDb(withObjectCount: 20)
        title = "Realm.io Bar Chart"
        
        _ = [["key": "toggleValues", "label": "Toggle Values"],
                       ["key": "toggleHighlight", "label": "Toggle Highlight"],
                       ["key": "animateX", "label": "Animate X"],
                       ["key": "animateY", "label": "Animate Y"],
                       ["key": "animateXY", "label": "Animate XY"],
                       ["key": "saveToGallery", "label": "Save to Camera Roll"],
                       ["key": "togglePinchZoom", "label": "Toggle PinchZoom"],
                       ["key": "toggleAutoScaleMinMax", "label": "Toggle auto scale min/max"]]
        
        
        chartView.delegate = self
        setupBarLineChartView(    chartView)
        setData()
    }

    func setData()
    {
        let realm = RLMRealm.default()
        let results = RealmDemoData.allObjects(in: realm)
        let set = RealmBarDataSet(results: results, xValueField: "xValue", yValueField: "yValue", label: "")
        
        set.colors = [ChartColorTemplates.colorFromString("#FF5722"), ChartColorTemplates.colorFromString("#03A9F4")]
        
        set.label = "Realm BarDataSet"
        let dataSets: [IChartDataSet] = [set]
        let data = BarChartData(dataSets: dataSets)
        styleData(data)
        
        chartView.fitBars = true
        chartView.data = data
        chartView.animate(yAxisDuration: 1.4, easingOption: .easeOutQuad)
    }

    func optionTapped( sender: NSMenuItem)
    {
        switch (sender.title)
        {
        default:
            super.toggle(sender.title, chartView: chartView)
        }
    }
}


// MARK: - ChartViewDelegate
extension RealmBarChartViewController: ChartViewDelegate
{
    public func chartValueSelected(_ chartView: ChartViewBase, entry: ChartDataEntry, highlight: Highlight)
    {
        print("chartValueSelected : x = \(highlight.x)")
    }
    
    public func chartValueNothingSelected(_ chartView: ChartViewBase)
    {
        print("chartValueNothingSelected")
    }
}

